File: src/com/aslan/module/cipher/segments/SerialSegmentDecipher.js
import AbstractSegment from "./AbstractSetment"

export default class SerialSegmentDecipher extends AbstractSegment {

	run1(IN, inf, length) {
		let algorithm = this.algorithm;
		let key = this.key;
		let L = this.info.L;
		let N = this.info.N;
		let round = length >> L >> 3;
		let remainder = length >> L & 7;
		for (let i = 0; i < round; i++) {
			algorithm.enc(new Uint8Array(IN, inf, N), key.update());
			inf += N;
			algorithm.enc(new Uint8Array(IN, inf, N), key.update());
			inf += N;
			algorithm.enc(new Uint8Array(IN, inf, N), key.update());
			inf += N;
			algorithm.enc(new Uint8Array(IN, inf, N), key.update());
			inf += N;
			algorithm.enc(new Uint8Array(IN, inf, N), key.update());
			inf += N;
			algorithm.enc(new Uint8Array(IN, inf, N), key.update());
			inf += N;
			algorithm.enc(new Uint8Array(IN, inf, N), key.update());
			inf += N;
			algorithm.enc(new Uint8Array(IN, inf, N), key.update());
			inf += N;
		}
		for (let i = 0; i < remainder; i++) {
			algorithm.enc(new Uint8Array(IN, inf, N), key.update());
			inf += N;
		}
		return length;
	}

	run2(IN, inf, OUT, ouf, length) {
		let algorithm = this.algorithm;
		let key = this.key;
		let N = this.info.N, L = this.info.L;
		let H = N - this.fill, F = N;
		let round = length >> L >> 2;
		let remainder = length >> L & 3;
		let buf;
		for (let i = 0; i < round; i++) {
			buf = new Uint8Array(IN, inf, N);
			algorithm.dec(buf, key.update());
			new Uint8Array(OUT, ouf, N).set(buf);
			inf += F;
			ouf += H;

			buf = new Uint8Array(IN, inf, N);
			algorithm.dec(buf, key.update());
			new Uint8Array(OUT, ouf, N).set(buf);
			inf += F;
			ouf += H;

			buf = new Uint8Array(IN, inf, N);
			algorithm.dec(buf, key.update());
			new Uint8Array(OUT, ouf, N).set(buf);
			inf += F;
			ouf += H;

			buf = new Uint8Array(IN, inf, N);
			algorithm.dec(buf, key.update());
			new Uint8Array(OUT, ouf, N).set(buf);
			inf += F;
			ouf += H;

		}
		for (let i = 0; i < remainder; i++) {
			buf = new Uint8Array(IN, inf, N);
			algorithm.dec(buf, key.update());
			new Uint8Array(OUT, ouf, N).set(buf);
			inf += F;
			ouf += H;
		}
		return Math.ceil(1.0 * length / F) * H;
	}
}

File: src/com/aslan/module/cipher/segments/SerialSegmentDecipher.java
package com.aslan.module.cipher.segments;

public class SerialSegmentDecipher extends AbstractSegment {

    public int exec(byte[] in, int inf, int length) {
        int L = info.L;
        int N = info.N;
        int round = length >> L >> 3;
        int remainder = length >> L & 7;
        for (int i = 0; i < round; i++) {
            algorithm.enc(in, inf, key.update());
            inf += N;
            algorithm.enc(in, inf, key.update());
            inf += N;
            algorithm.enc(in, inf, key.update());
            inf += N;
            algorithm.enc(in, inf, key.update());
            inf += N;
            algorithm.enc(in, inf, key.update());
            inf += N;
            algorithm.enc(in, inf, key.update());
            inf += N;
            algorithm.enc(in, inf, key.update());
            inf += N;
            algorithm.enc(in, inf, key.update());
            inf += N;
        }
        for (int i = 0; i < remainder; i++) {
            algorithm.enc(in, inf, key.update());
            inf += N;
        }
        return length;
    }

    public int exec(byte[] in, int inf, byte[] out, int ouf, int length) {
        int N = info.N, L = info.L;
        int H = N - fill, F = N;
        int round = length >> L >> 2;
        int remainder = length >> L & 3;
        for (int i = 0; i < round; i++) {
            algorithm.dec(in, inf, key.update());
            System.arraycopy(in, inf, out, ouf, H);
            inf += F;
            ouf += H;
            algorithm.dec(in, inf, key.update());
            System.arraycopy(in, inf, out, ouf, H);
            inf += F;
            ouf += H;
            algorithm.dec(in, inf, key.update());
            System.arraycopy(in, inf, out, ouf, H);
            inf += F;
            ouf += H;
            algorithm.dec(in, inf, key.update());
            System.arraycopy(in, inf, out, ouf, H);
            inf += F;
            ouf += H;
        }
        for (int i = 0; i < remainder; i++) {
            algorithm.dec(in, inf, key.update());
            System.arraycopy(in, inf, out, ouf, H);
            inf += F;
            ouf += H;
        }
        return (int) (Math.ceil(1.0 * length / F) * H);
    }
}